prov/tcp/src/tcpx_comm.c

```diff
@@ -36,135 +36,122 @@
-	if (ofi_rbempty(&pe_entry->comm_buf)) {
-		if (len <= pe_entry->cache_sz) {
-			tcpx_comm_recv_buffer(pe_entry);
-		} else {
-			return tcpx_comm_recv_socket(pe_entry->ep->conn_fd, buf, len);
-		}
+	size_t len = 0;
+	uint64_t buf_offset;
+	int iov_offset, i;
+
+	prepare_iovec(iovec, pe_entry);
+
+	iov_offset = 0;
+	buf_offset = pe_entry->done_len;
+	for (i = 0 ; i < pe_entry->msg_data.iov_cnt+1 ; i++) {
+		len += iovec[i].iov_len;
+		if (pe_entry->done_len < len)
+			continue;
+
+		iov_offset = i+1;
+		buf_offset = pe_entry->done_len - len;
 	}
 
-	read_len = MIN(len, ofi_rbused(&pe_entry->comm_buf));
-	ofi_rbread(&pe_entry->comm_buf, buf, read_len);
-	FI_DBG(&tcpx_prov, FI_LOG_EP_DATA, "read from buffer: %lu\n", read_len);
-	return read_len;
+	iovec[iov_offset].iov_base = ((uint8_t *)(iovec[iov_offset].iov_base) +
+				      buf_offset);
+
+	iovec[iov_offset].iov_len -= buf_offset;
+
+	msghdr->msg_name = NULL;
+	msghdr->msg_namelen = 0;
+	msghdr->msg_iov = &iovec[iov_offset];
+	msghdr->msg_iovlen = pe_entry->msg_data.iov_cnt+1 - iov_offset;
+	msghdr->msg_control = NULL;
+	msghdr->msg_controllen = 0;
+	msghdr->msg_flags = 0;
 }
 
-static ssize_t tcpx_comm_send_socket(SOCKET conn_fd, const void *buf, size_t len)
+int tcpx_send_msg(struct tcpx_pe_entry *pe_entry)
 {
-	ssize_t ret;
+	struct msghdr msg;
+	struct iovec iovec[TCPX_IOV_LIMIT+1];
+	ssize_t bytes_sent;
 
-	ret = ofi_send_socket(conn_fd, buf, len, MSG_NOSIGNAL);
-	if (ret >= 0) {
-		FI_DBG(&tcpx_prov, FI_LOG_EP_DATA, "wrote to network: %lu\n", ret);
-		return ret;
-	}
+	prepare_msg(&msg, iovec, pe_entry);
 
-	if (OFI_SOCK_TRY_SND_RCV_AGAIN(ofi_sockerr())) {
-		ret = 0;
-	} else {
-		FI_DBG(&tcpx_prov, FI_LOG_EP_DATA,
-		       "write error: %s\n", strerror(ofi_sockerr()));
-	}
-	return ret;
+	bytes_sent = ofi_sendmsg_socket(pe_entry->ep->conn_fd, &msg, 0);
+	if (bytes_sent < 0)
+		return -errno;
+
+	pe_entry->done_len += bytes_sent;
+	return FI_SUCCESS;
 }
 
+static int tcpx_recv_msg_hdr(struct tcpx_pe_entry *pe_entry)
 
-ssize_t tcpx_comm_flush(struct tcpx_pe_entry *pe_entry)
 {
-	ssize_t ret1, ret2 = 0;
-	size_t endlen, len, xfer_len;
-
-	len = ofi_rbused(&pe_entry->comm_buf);
-	endlen = pe_entry->comm_buf.size -
-		(pe_entry->comm_buf.rcnt & pe_entry->comm_buf.size_mask);
-
-	xfer_len = MIN(len, endlen);
-	ret1 = tcpx_comm_send_socket(pe_entry->ep->conn_fd, (char*)pe_entry->comm_buf.buf +
-				     (pe_entry->comm_buf.rcnt & pe_entry->comm_buf.size_mask),
-				     xfer_len);
-	if (ret1 > 0)
-		pe_entry->comm_buf.rcnt += ret1;
-
-	if (ret1 == xfer_len && xfer_len < len) {
-		ret2 = tcpx_comm_send_socket(pe_entry->ep->conn_fd, (char*)pe_entry->comm_buf.buf +
-					     (pe_entry->comm_buf.rcnt & pe_entry->comm_buf.size_mask),
-					     len - xfer_len);
-		if (ret2 > 0)
-			pe_entry->comm_buf.rcnt += ret2;
-		else
-			ret2 = 0;
+	ssize_t bytes_recvd;
+	void *rem_hdr_buf;
+	size_t rem_hdr_len;
+
+	if (pe_entry->done_len >= sizeof(pe_entry->msg_hdr))
+		return FI_SUCCESS;
```

Comment:
It would be better to have this sort of check outside of the function, so we avoid the function call overhead completely.